### ISSUE TITLE
v4.2.0, Qt Scaling issues

### ISSUE THREAD

**Issue by dallenwilson** [Yubico/yubioath-desktop#213]
I'd like to preface this by saying I do not believe this is a bug in yubioath-desktop. It's probably not a bug in Qt, either, but a configuration/setting issue somewhere on my system.

I updated from v3.1.0 to v4.2.0 a few days ago, and the UI is nearly unusable. It can't be resized, the text/menus/fontsize is all very tiny. The behaviour changes a bit depending on if the window is on my primary monitor or not (The scaling is much, much smaller on secondary monitors). I've attached screenshots showing what I am seeing.

![primary](https://user-images.githubusercontent.com/1279015/33284338-5a0727c8-d374-11e7-9a4a-e0b34641af3e.png)
![non-primary](https://user-images.githubusercontent.com/1279015/33284473-db8520a2-d374-11e7-90f5-046c5e34a6b2.png)

This is on a Gentoo box with Cinnamon v3.0.7 as the WM. Yubioath-desktop was installed via Gentoo's package management system, portage. It pulled in the various qt-related dependencies automatically, and I haven't edited any Qt-related configuration files.

Can anyone point me in the right direction as far as fixing this goes?

**Comment by emlun** [comment]
Oh wow, indeed I see how that is unusable. This looks related to #203.

That looks to me like it's even smaller than it would be on a HiDPI monitor with no scaling. Somehow it feels like perhaps some scaling layer somewhere in the stack has the scaling factor backwards, and accidentally scales it to smaller instead of larger. Or perhaps we have to explicitly opt in to HiDPI scaling in Qt somehow - I'll have to read up on that.

Questions:

- Are all your monitors scaled, or just some of them?
- What happens if you switch around which monitor is the primary one?
- Have you had these issues with any other Qt5 apps?

**Comment by dallenwilson** [comment]
#203 rang a bell with me as well, but in the end I decided to open a new issue rather than tack this on to an existing one.

- I don't believe any of them are scaled. I'm not sure what scaling means in this context. I haven't explicitly set any scaling factors, through Qt or otherwise. Cinnamon's 'Text Scaling Factor' is set to 1.0, I haven't changed that either. I just installed qt5ct to check for any scaling/sizing options, but I don't see anything listed. It did warn me that an environment variable "QT_STYLE_OVERRIDE" was set to "gtk" and it had to be unset before continuing. I'm not sure where or why that is getting set in my profile, I'll have to dig in to that. Unsetting it prior to launching yubioath-desktop has no effect.
- Changing which is the primary doesn't affect anything, other than moving my menu bars / clock around.
- I don't use a lot of Qt apps to be honest, so I didn't have another one handy. After a quick search I found that Calibre, the eBook organizer/manager, also uses Qt5. I installed it and am seeing the same behaviour. The only three differences being that I can manually resize the Calibre's window by clicking and dragging the corners, everything is usable (if anything, a bit too large) on my primary monitor, and moving it to a secondary results in the same tiny, tiny UI as yubioath-desktop.

It may be worth mentioning here that these are three different models of monitor, from three different manufacturers. I have each set to it's max supported resolution (which again are three different numbers), which you can see from the xrandr displayed in the screenshots. They're all plugged in to the same physical video card, an Nvidia 750Ti, and I'm running the nvidia drivers, not nouveau or mesa.

**Comment by emlun** [comment]
Oh, sorry, I didn't notice the xrandr dumps and jumped to the conclusion that this was an issue with HiDPI (i.e., 4K) displays scaled to lower resolutions.

The fact that you see the same issue in Calibre could mean that this is a Qt or Cinnamon issue, but it also doesn't rule out that both Yubico Authenticator and Calibre do something wrong in the same way. On the other hand I don't see these issues in i3wm. I'll see if I can reproduce the behaviour in Cinnamon or Gnome Shell.

**Comment by dallenwilson** [comment]
I spent a few hours looking in to it before opening an issue here, and I'm almost certain it's a configuration issue on my system, rather than a bug in the application, or even in Qt. I wasn't able to find any records of similar issues from anyone else using Qt5-based applications. Actually I couldn't find much in relation to Qt5 at all. I opened the issue hoping you'd have some suggestions as far as env flags to set, config options that can be changed, etc. And to give Google something to find for the next one with an issue like this.

I'll keep digging in to it, I don't have a lot of experience with Qt-based anything, but I'll find out where to go looking for more information. Logfiles, verbose flags I can turn on, etc. I'm sure there's evidence somewhere, I just need to find it.

**Comment by emlun** [comment]
Sorry, I'm afraid I don't know either. All I can say is that I didn't manage to reproduce the issue in either Cinnamon or Gnome Shell on an Ubuntu VM (I tried to test in Gentoo as well, but the machine crashed 2/3-way through emerging the ~300 packages...).

Perhaps @dagheyman can add something when he returns from his vacation.

**Comment by dallenwilson** [comment]
No worries. Probably safe to close this then, for the moment. I'll re-open and follow up if/when I figure out what's going on here.

**Comment by dallenwilson** [comment]
It's all sorted out now. My primary monitor (A Samsung SyncMaster 2494HM, attached via HDMI) is reporting bad EDID information. It's a 24" monitor, and Xorg/nvidia-proprietary-drivers/xrandr/get-edid/etc-and-so-on are all identifying it as a 7" monitor, which in turn is throwing the nvidia-drivers DPI calculation completely off kilter.

Qt5 then takes that wrong DPI information, attempts to scale things to look correct on the primary monitor, which results in teeny-tiny-everything on the others. Manually specifying the correct monitor information in /etc/X11/xorg.conf did not work, apparently Nvidia's proprietary driver can be a bit obstinate about having it's EDID auto-scan overridden. To be fair, it is supposed to be possible to do. I only gave it three or four attempts before I got annoyed and went digging for information on how to override Qt5's auto-scale stuff instead.

In the end, my fix was adding the following to my ~/.bashrc:
export "QT_AUTO_SCREEN_SCALE_FACTOR=0" #Disable auto scaling entirely
export "QT_SCREEN_SCALE_FACTORS=1;1;1" #Specify scale factor of 1 for all three monitors
export "QT_SCALE_FACTOR=1" #Set monitor-independent scale factor to 1

The result is an entirely usable Qt5 application, on any of my three monitors. Of course, yubioath-desktop does not appear exactly the same size as it moves from one to another, but since that is what I expect from multiple monitors at different resolutions, I am very ok with this result!
